google/cloud/pubsub/internal/tracing_exactly_once_ack_handler.cc

```diff
@@ -0,0 +1,150 @@
+
+  ~TracingExactlyOnceAckHandler() override = default;
+  future<Status> ack() override {
+    if (subscribe_span_.has_value()) {
+      subscribe_span_.value()->AddEvent("gl-cpp.message_ack");
+    }
+    namespace sc = opentelemetry::trace::SemanticConventions;
+    opentelemetry::trace::StartSpanOptions options = RootStartSpanOptions();
+
+    options.kind = opentelemetry::trace::SpanKind::kInternal;
+    auto const ack_id = child_->ack_id();
+    auto const subscription = child_->subscription();
+    TracingAttributes attributes = MakeSharedAttributes(ack_id, subscription);
+    attributes.emplace_back(
+        std::make_pair(sc::kCodeFunction, "pubsub::AckHandler::ack"));
+    auto span = internal::MakeSpan(subscription.subscription_id() + " ack",
+                                   attributes, options);
+
+    auto scope = internal::OTelScope(span);
+    return child_->ack().then(
+        [oc = opentelemetry::context::RuntimeContext::GetCurrent(),
+         span = std::move(span)](auto f) {
+          auto result = f.get();
+          internal::DetachOTelContext(oc);
+          return internal::EndSpan(*span, std::move(result));
+        });
+  }
+
+  future<Status> nack() override {
+    if (subscribe_span_.has_value()) {
+      subscribe_span_.value()->AddEvent("gl-cpp.message_nack");
+    }
+    namespace sc = opentelemetry::trace::SemanticConventions;
+    opentelemetry::trace::StartSpanOptions options = RootStartSpanOptions();
+
+    options.kind = opentelemetry::trace::SpanKind::kInternal;
+    auto const ack_id = child_->ack_id();
+    auto const subscription = child_->subscription();
+    TracingAttributes attributes = MakeSharedAttributes(ack_id, subscription);
+    attributes.emplace_back(
+        std::make_pair(sc::kCodeFunction, "pubsub::AckHandler::nack"));
+    auto span = internal::MakeSpan(subscription.subscription_id() + " nack",
+                                   attributes, options);
+
+    auto scope = internal::OTelScope(span);
+    return child_->nack().then(
+        [oc = opentelemetry::context::RuntimeContext::GetCurrent(),
+         span = std::move(span)](auto f) {
+          auto result = f.get();
+          internal::DetachOTelContext(oc);
+          return internal::EndSpan(*span, std::move(result));
+        });
+  }
+
+  std::int32_t delivery_attempt() const override {
+    return child_->delivery_attempt();
+  }
+
+  std::string ack_id() override { return child_->ack_id(); }
+
+  pubsub::Subscription subscription() const override {
+    return child_->subscription();
+  }
+
+ private:
+  TracingAttributes MakeSharedAttributes(
+      std::string const& ack_id, pubsub::Subscription const& subscription) {
+    namespace sc = opentelemetry::trace::SemanticConventions;
+    return TracingAttributes{
+        {sc::kMessagingSystem, "gcp_pubsub"},
+        {"messaging.gcp_pubsub.message.ack_id", ack_id},
+        {"messaging.gcp_pubsub.subscription.template", subscription.FullName()},
```

Comment:
The second element of these pairs is a `opentelemetry::common::AttributeValue`, which only holds a `opentelemetry::nostd::string_view`, and that is left dangling when the `std::string` returned by `subscription.FullName()` is destroyed.

Reply:
Thanks Brad. I'm not sure if my change fixes it 

```
    std::string(subscription.FullName())},
```

Reply:
> I'm not sure if my change fixes it

A successful `asan-pr` will give you some confidence.

Reply:
I don't think it will make any difference ... the result of `FullName()` is already a `std::string`.

The problem is much more fundamental.  As things stand, `TracingAttributes` is a referencing type, so you have to be careful with lifetimes when using instances.  Returning such a type from a function is typically fragile.

Reply:
abandoning the helper, and just spelling it out twice seems reasonable.

Reply:
Even then you'll undoubtedly need to give `subscription.FullName()` a name, and ensure that it lives longer than the `TracingAttributes`.

Reply:
> ensure that it lives longer

When a span is created, the exporter will make owning copies of the attribute `string_view`s. So we only need something that lives through our call to `internal::MakeSpan(...)`

Reply:
OK ... as long as the string copy is made before the demise of the viewed strings, all should be well.

You might consider, however, making the "attributes" type do the copying, and then consuming (moving) those strings when creating the span.

Reply:
The new code passes because the temporary string isn't destroyed until the end of the statement, by which time it has already been copied (via the `string_view`).  That feels fragile, but OK.

Reply:
ack